### ISSUE TITLE
Write API does not return success or error result

### ISSUE THREAD

**Issue by RalfMengwasser** [influxdata/influxdb-client-csharp#107]
The whole write API does not return with a result and instead just returns Task or void.

Instead the API should return the HTTP status or an aggregated HTTP status in case of batched writes so that we can immediately know if the write succeeded, as it is very important data does not get lost.

The current workaround would be to hook into the one event there is. There aren't even proper events to distinguish between success or errors, which means I have to check every time which one it is.

And even if I do that, as I may have lots of different threads doing writes, how would I know to which write the callback belongs? There doesn't seem to be any tag or something for me to match the event callback to the write command.

This API is really hard to use as the functions aren't really useful.

**Comment by bednar** [comment]
Hi @RalfMengwasser,

thanks for feedback. 

The `WriteApi` writes data in background thread so we aren't able to return aggregated HTTP status, but you could hook to errors by:

```c#
writeApi.EventHandler += (sender, eventArgs) =>
{
    if (eventArgs is WriteErrorEvent @event)
    {
        var exception = @event.Exception;
        
        //
        // handle error
        //
    }
};
```
https://github.com/influxdata/influxdb-client-csharp/blob/d5c0e544a87a9a7118ccdb1c55db650e7c5c8520/Client/README.md#handle-the-error-write

The `WriteApi` produces batches that are configured by: `BatchSize` and `FlushInterval`. This batches could aggregate data from different threads so you are not able to know which write belongs to concrete thread.

On the other hand the `WriteApiAsync` writes data directly into InfluxDB.  If there is no an exception then data are successfully written, but if there is a something wrong, the exception contains all information:

```csharp
try
{
    await _writeApi.WriteRecordAsync(WritePrecision.S, "h2o,location=coyote_creek 1");
}
catch (Exception e)
{
    Console.WriteLine(e);
}
```

![Screenshot 2020-07-03 at 07 58 24](https://user-images.githubusercontent.com/455137/86436658-fd91da80-bd02-11ea-8d57-0349d600f365.png)

**Comment by RalfMengwasser** [comment]
> The WriteApi writes data in background thread so we aren't able to return aggregated HTTP status, but you could hook to errors by:
>The WriteApi produces batches that are configured by: BatchSize and FlushInterval. This batches could aggregate data from different threads so you are not able to know which write belongs to concrete thread.

There is no way for me to tell WHICH data did not get saved, so any errors are useless.

An example:

In my case I have a list of datapoint batches that need to be saved. RELIABLY.

Batch 1 has 5000 points.
Batch 2 has 5000 points.

I call Write on Batch 1. I call Write on Batch 2.

Exception returned.

Did it fail on Batch 1 or Batch 2? Or did it mix the batches? Did it fail on 50% of batch 1 and 50% of batch?

I have another multi-input cache "database" in front of InfluxDB that hands of these datapoints, and not even a single one of these points can get lost in the process. So I need to get confirmation that these are STORED in InfluxDB before I delete them from the cache database.

I need functions that tell me WHICH of these points I put into the call are actually saved and which are not. In fact, ANY database API I have ever seen will tell you if the write call you made was successful.

So what do I do now? In case of ANY exception in X seconds I rewrite ALL batches that are in the process of being saved?

Also, did you notice that most of the logic in the WriteAPI is actually in the CONSTRUCTOR? What happened there? Who wrote that code? I was thinking of actually changing the code to do what is needed, but when I looked at it it was clear that this needs to be rewritten from scratch as the code is a complete mess.

Thank god the Async API seems to be written in a much more maintainable way, so I will probably go there and fix the shortcomings.

But holy god.

**Comment by slyce130** [comment]
@RalfMengwasser Do you have experience with reactive extensions? I think it makes sense that WriteApi subscribes to the observable to make it hot the moment its created. I suppose the logic itself could be moved elsewhere, but I don't see anything fundamentally wrong that it would need to be rewritten from scratch.

If you can explain why you think this is wrong, I'd find it insightful.

**Comment by Ahrimaan** [comment]
> 
> 
> @RalfMengwasser Do you have experience with reactive extensions? I think it makes sense that WriteApi subscribes to the observable to make it hot the moment its created. I suppose the logic itself could be moved elsewhere, but I don't see anything fundamentally wrong that it would need to be rewritten from scratch.
> 
> If you can explain why you think this is wrong, I'd find it insightful.

So you always have to put some infos int a event to find out where the write did not happened. Thats why a clear return code is always better than void and wait for some kind of events.

**Comment by bednar** [comment]
The issue is fixed in [1.19.0](https://github.com/influxdata/influxdb-client-csharp/milestone/22) milestone. You can use latest dev version by:

```console
dotnet add package InfluxDB.Client --version 1.19.0-dev.3051
```

[`WriteApiAsync`](https://github.com/influxdata/influxdb-client-csharp/blob/6636bec115c7d751c02681571e814423f87f0e53/Client/WriteApiAsync.cs#L15) now supports return `IRestResponse` for custom handling of HTTP response:

- https://github.com/influxdata/influxdb-client-csharp/blob/6636bec115c7d751c02681571e814423f87f0e53/Client/WriteApiAsync.cs#L185
- https://github.com/influxdata/influxdb-client-csharp/blob/6636bec115c7d751c02681571e814423f87f0e53/Client/WriteApiAsync.cs#L314
- https://github.com/influxdata/influxdb-client-csharp/blob/6636bec115c7d751c02681571e814423f87f0e53/Client/WriteApiAsync.cs#L484
